pjlib/src/pjlib-test/stack.c

```diff
@@ -0,0 +1,389 @@
+static int worker_thread(void* p);
+
+static stack_test_desc tests[] = {
+    {
+        .cfg.title = "stack (single thread)",
+        .cfg.n_threads = 0,
+        .cfg.repeat = MAX_REPEATS
+    },
+    {
+        .cfg.title = "stack (multi threads)",
+        .cfg.n_threads = MAX_THREADS,
+        .cfg.repeat = MAX_REPEATS
+    }
+};
+
+int stack_test()
+{
+    pj_stack_type   *stack = NULL;
+    const int        sz = 15;
+    stack_node      *nodes;
+    stack_node      *p;
+    int i; // don't change to unsigned!
+
+    pj_pool_t       *pool = NULL;;
+    pj_status_t      status;
+    int              rc;
+
+    pool = pj_pool_create(mem, NULL, 4096, 0, NULL);
+    if (!pool)
+    {
+        rc = -10;
+        goto error;
+    }
+
+    nodes = pj_pool_calloc(pool, sz, sizeof(stack_node));
+    if (!nodes)
+    {
+        rc = -15;
+        goto error;
+    }
+
+#if defined(PJ_WIN32) && PJ_STACK_IMPLEMENTATION==PJ_STACK_WIN32
+
+    /* Here we check our alignment macros PJ_STACK_ALIGN_PREFIX, PJ_STACK_ALIGN_SUFFIX */
+
+#   if  defined(_MSC_VER)
+#       pragma warning(push)                                                  
+#       pragma warning(disable:4324)   // structure padded due to align()
+#   endif  // defined(_MSC_VER)
+
+    if (TYPE_ALIGNMENT(stack_node) < MEMORY_ALLOCATION_ALIGNMENT)
+    {
+        //#error alignment of array's items used as stack node under Windows platform should not be less than MEMORY_ALLOCATION_ALIGNMENT
+        rc = -3;
+        goto error;
+    }
+
+#   if  defined(_MSC_VER)
+#       pragma warning(pop)
+#   endif  // defined(_MSC_VER)
+
+    for (i = 0; i < sz; ++i) {
+        if (!IS_ALIGNED_PTR(&nodes[i], MEMORY_ALLOCATION_ALIGNMENT))
+        {
+            rc = -7;
+            goto error;
+        }
+    }
+
+#endif  // PJ_STACK_IMPLEMENTATION==PJ_STACK_WIN32
+
+    //const char* stack_test = "stack_test";
+    // test stack_create()
+    if ((status = pj_stack_create(pool, /*stack_test,*/ &stack)) != PJ_SUCCESS)
+    {
+        rc = -20;
+        goto error;
+    }
+
+    // created stack is empty
+    if (pj_stack_pop(stack))
+    {
+        rc = -25;
+        goto error;
+    }
+
+    // Test push().
+    for (i = 0; i < sz; ++i) {
+        nodes[i].value = i;
+        if ((status= pj_stack_push(stack, &nodes[i])) != PJ_SUCCESS)
+        {
+            rc = -30;
+            goto error;
+        }
+    }
+
+    // test pop().
+    while((p = pj_stack_pop(stack)) != NULL) {
+        --i;
+        pj_assert(p->value == i); //FILO !
+        if (p->value != i || p != nodes + i) {  
+            rc = -40;
+            goto error;
+        }
+    }
+    if (i)
+    {
+        rc = -50;
+        goto error;
+    }
+    rc = 0;
+error:
+    if (stack != NULL && pj_stack_destroy(stack) != PJ_SUCCESS && !rc)
+    {
+        rc = -55;
+    }
+
+    for (i = 0; !rc && i < PJ_ARRAY_SIZE(tests); ++i) {
+        tests[i].state.pool = pool;
+        rc = stack_stress_test(&tests[i]);
```

Comment:
perhaps we should by default disable the stress test for non-Windows platform (create a macro such as #HAS_STRESS_TEST PJ_WIN32).

It will only burden the CI machine and there's little point in stress testing a relatively slower implementation of LIFO list + mutex.

Reply:
I think we need multithreaded testing for thread safe api, but currently the default implementation is cross platform so may be just Windows testing  is enough.
Create a macro HAS_MT_STACK_STRESS_TEST

all fixed!

> These should be all from me

I hope this will be the case this time! :)